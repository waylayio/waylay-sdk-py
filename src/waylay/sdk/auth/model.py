"""Utilities to handle waylay authentication."""

from datetime import datetime
from enum import Enum
from dataclasses import dataclass
from typing import Optional, ClassVar, Dict, Any, List
import json
import abc

import base64
import binascii
from jose import jwt, JWTError

from .exceptions import TokenParseError, AuthError


class CredentialsType(str, Enum):
    """Supported Waylay Authentication Methods.

    Note that username/password authentication (as used in our IDP at
    https://login.waylay.io)
    is not (yet) supported.

    """

    CLIENT = "client_credentials"
    APPLICATION = "application_credentials"
    TOKEN = "token"
    CALLBACK = "interactive"

    def __str__(self):
        """Get the string representation."""
        return self.value


TokenString = str


class WaylayCredentials(abc.ABC):
    """Base class for the representation of credentials to the waylay platform."""

    gateway_url: Optional[str] = None
    credentials_type: ClassVar[CredentialsType] = CredentialsType.CALLBACK

    # legacy
    accounts_url: Optional[str] = None

    @abc.abstractmethod
    def to_dict(self, obfuscate=True) -> Dict[str, Any]:
        """Convert the credentials to a json-serialisable representation."""

    @abc.abstractproperty
    def id(self) -> Optional[str]:
        """Get the main identifier for this credential."""
        return None

    def __repr__(self):
        """Show the implementing class and public information."""
        return f"<{self.__class__.__name__}({str(self)})>"

    def __str__(self):
        """Show the credential attributes, with secrets obfuscated."""
        return json.dumps(self.to_dict(obfuscate=True))

    @abc.abstractmethod
    def is_well_formed(self) -> bool:
        """Validate that these credentials are well-formed.

        This does not assure that they will lead to a succesfull
        authentication.

        """


@dataclass(repr=False)
class AccountsUrlMixin:
    """Dataclass mixin for the 'gateway_url' (legacy 'accounts_url') property."""

    gateway_url: Optional[str] = None
    accounts_url: Optional[str] = None


@dataclass(repr=False, init=False)
class ApiKeySecretMixin(AccountsUrlMixin):
    """Dataclass mixin for the 'api_key' and 'api_secret'."""

    api_key: str = ""
    api_secret: str = ""

    def __init__(
        self,
        api_key: str,
        api_secret: str,
        *,
        gateway_url: Optional[str] = None,
        accounts_url: Optional[str] = None,
    ):
        """Initialise with the api_key and api_secret."""
        super().__init__(gateway_url=gateway_url, accounts_url=accounts_url)
        self.api_key = api_key
        self.api_secret = api_secret

    @property
    def id(self) -> Optional[str]:
        """Get the main identifier for this credential."""
        return self.api_key

    @classmethod
    def create(
        cls,
        api_key: str,
        api_secret: str,
        *,
        gateway_url: Optional[str] = None,
        accounts_url: Optional[str] = None,
    ):
        """Create a client credentials object."""
        return cls(
            api_key=api_key,
            api_secret=api_secret,
            gateway_url=gateway_url,
            accounts_url=accounts_url,
        )

    def to_dict(self, obfuscate=True):
        """Convert the credentials to a json-serialisable representation."""
        return dict(
            type=self.credentials_type.value,
            api_key=self.api_key,
            api_secret="********" if obfuscate else self.api_secret,
            gateway_url=self.gateway_url,
            accounts_url=self.accounts_url,
        )

    def is_well_formed(self) -> bool:
        """Validate that these credentials are well-formed.

        This does not assure that they will lead to a succesfull
        authentication.

        """
        if not (self.api_key and self.api_secret):
            return False
        # api key are 12 bytes hexc encoded
        try:
            if len(bytes.fromhex(self.api_key)) != 12:
                return False
        except ValueError:
            return False
        # api secret are 24 bytes base64 encoded (rfc4648)
        try:
            if len(base64.b64decode(self.api_secret, validate=True)) != 24:
                return False
        except binascii.Error:
            return False
        return True


@dataclass(repr=False, init=False)
class NoCredentials(AccountsUrlMixin, WaylayCredentials):
    """Represents credentials which be asked via (interactive) callback when required."""

    credentials_type: ClassVar[CredentialsType] = CredentialsType.CALLBACK

    def to_dict(self, obfuscate=True):  # pylint: disable=unused-argument
        """Convert the credentials to a json-serialisable representation."""
        return dict(
            type=str(self.credentials_type),
            gateway_url=self.gateway_url,
            accounts_url=self.accounts_url,
        )

    def is_well_formed(self) -> bool:
        """Validate that these credentials are well-formed."""
        return True

    @property
    def id(self) -> Optional[str]:
        """Get the main identifier for this credential."""
        return None


@dataclass(repr=False, init=False)
class ClientCredentials(ApiKeySecretMixin, WaylayCredentials):
    """Waylay Credentials: api key and secret of type 'client_credentials'."""

    credentials_type: ClassVar[CredentialsType] = CredentialsType.CLIENT


@dataclass(repr=False, init=False)
class ApplicationCredentials(ApiKeySecretMixin, WaylayCredentials):
    """Waylay Credentials: api key and secret of type 'application_credentials'."""

    credentials_type: ClassVar[CredentialsType] = CredentialsType.APPLICATION
    tenant_id: str = ""


@dataclass(repr=False, init=False)
class TokenCredentials(AccountsUrlMixin, WaylayCredentials):
    """Waylay JWT Token credentials."""

    credentials_type: ClassVar[CredentialsType] = CredentialsType.TOKEN
    token: TokenString = ""

    def __init__(
        self,
        token: TokenString,
        *,
        gateway_url: Optional[str] = None,
        accounts_url: Optional[str] = None,
    ):
        """Create a TokenCredentials from a token string."""
        super().__init__(gateway_url=gateway_url, accounts_url=accounts_url)
        self.token = token

    @property
    def id(self) -> Optional[str]:
        """Get the main identifier for this credential."""
        try:
            token = WaylayToken(self.token)
            return f"[{token.domain}] {token.subject}"
        except AuthError as exc:
            return f"INVALID_TOKEN({exc})"

    def to_dict(self, obfuscate=True):
        """Get the credential attributes."""
        return dict(
            type=str(self.credentials_type),
            token="*********" if obfuscate else self.token,
            gateway_url=self.gateway_url,
            accounts_url=self.accounts_url,
        )

    def is_well_formed(self) -> bool:
        """Validate that these credentials are well-formed."""
        try:
            # WaylayToken constructor decodes the data without signature verification
            return WaylayToken(self.token).tenant is not None
        except TokenParseError:
            return False


class WaylayToken:
    """Holds a Waylay JWT token."""

    def __init__(self, token_string: str, token_data: Optional[Dict] = None):
        """Create a Waylay Token holder object from given token string or data."""
        self.token_string = token_string
        if token_data is None:
            try:
                token_data = jwt.decode(
                    token_string, "", options=dict(verify_signature=False)
                )
            except (TypeError, ValueError, JWTError) as exc:
                raise TokenParseError(exc) from exc
        self.token_data = token_data

    def validate(self) -> "WaylayToken":
        """Verify essential assertions, and its expiry state.

        This implementation does not verify the signature of a token, as
        this is seen the responsability of a server implementation.

        """
        if not self.token_string:
            raise AuthError("no token")

        if not self.token_data:
            raise AuthError("could not parse token data")

        if not self.tenant:
            raise AuthError("invalid token")

        # assert expiry
        if self.is_expired:
            raise AuthError("token expired")
        return self

    @property
    def tenant(self) -> Optional[str]:
        """Get the tenant id asserted by the token."""
        return self.token_data.get("tenant", None)

    @property
    def domain(self) -> Optional[str]:
        """Get the waylay domain asserted by the token."""
        return self.token_data.get("domain", None)

    @property
    def subject(self) -> Optional[str]:
        """Get the subject asserted by the token."""
        return self.token_data.get("sub", None)

    @property
    def licenses(self) -> List[str]:
        """Get the licenses asserted by the token."""
        return self.token_data.get("licenses", [])

    @property
    def groups(self) -> List[str]:
        """Get the groups asserted by the token."""
        return self.token_data.get("groups", [])

    @property
    def permissions(self) -> List[str]:
        """Get the permissions asserted by the token."""
        return self.token_data.get("permissions", [])

    @property
    def expires_at(self) -> Optional[datetime]:
        """Get the token expiry timestamp."""
        exp = self.token_data.get("exp", None)
        return None if exp is None else datetime.fromtimestamp(exp)

    @property
    def issued_at(self) -> Optional[datetime]:
        """Get the token issuance timestamp."""
        iat = self.token_data.get("iat", None)
        return None if iat is None else datetime.fromtimestamp(iat)

    @property
    def expires_seconds(self) -> int:
        """Get seconds until expiry."""
        exp = self.token_data.get("exp", None)
        return 0 if exp is None else exp - datetime.now().timestamp()

    @property
    def age(self) -> int:
        """Get seconds sinds issuance."""
        iat = self.token_data.get("iat", 0)
        return int(datetime.now().timestamp() - iat)

    @property
    def is_expired(self) -> bool:
        """Get the expiration state.

        True if a (previously valid) the token has expired.

        """
        if not isinstance(self.token_data, dict):
            return True
        exp = self.expires_at
        return exp is None or exp < datetime.now()

    @property
    def is_valid(self) -> bool:
        """Get the token validity.

        True if essential token data is present and is not expired.

        """
        return (
            self.tenant is not None
            and self.subject is not None
            and self.domain is not None
            and not self.is_expired
        )

    def to_dict(self):
        """Get the main token attributes."""
        return dict(
            tenant=self.tenant,
            domain=self.domain,
            subject=self.subject,
            expires_at=str(self.expires_at),
            is_valid=self.is_valid,
        )

    def __repr__(self) -> str:
        """Show the implementing class an main attributes."""
        return f"<{self.__class__.__name__}({json.dumps(self.to_dict())})>"

    def __str__(self) -> str:
        """Render the token string."""
        return self.token_string

    def __bool__(self) -> bool:
        """Get the validity of the token."""
        return self.is_valid
