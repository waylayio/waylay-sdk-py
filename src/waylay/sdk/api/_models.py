from __future__ import annotations
from abc import ABC
from datetime import date, datetime
from decimal import Decimal
from inspect import isclass
from typing import TYPE_CHECKING, Any, Callable, Dict, List, Union, get_type_hints

from typing_extensions import (
    Annotated,  # >=3.9
    Self,  # >=3.12
    TypeAliasType,  # >=3.12
)

if TYPE_CHECKING:
    from typing_extensions import TypeAlias  # >= Python 3.10

from pydantic import (
    BaseModel as PydanticBaseModel,
    ConfigDict,
    SerializationInfo,
    StrictStr,
    TypeAdapter,
    ValidationError,
    ValidationInfo,
    ValidatorFunctionWrapHandler,
    field_validator,
    model_serializer,
    model_validator,
)
from pydantic.functional_validators import ModelWrapValidatorHandler


class BaseModel(PydanticBaseModel, ABC):
    """Waylay base model class that adds some additional methods to Pydantic's `BaseModel`, including a custom validator and serializer."""

    @model_serializer(mode="wrap")
    def _model_serializer(
        self, handler: Callable, info: SerializationInfo
    ) -> Dict[StrictStr, Any]:
        """The default serializer of the model.

        * Excludes `None` fields that were not set at model initialization.
        """
        model_dict = handler(self, info)
        return {
            k: v
            for k, v in model_dict.items()
            if v is not None or k in self.model_fields_set
        }

    @model_validator(mode="wrap")  # type: ignore[arg-type]
    def _model_validator(
        cls, value: Any, handler: ModelWrapValidatorHandler, info: ValidationInfo
    ):
        """The default validator of the model.

        When validation is called with a `skip_validation=True` context (e.g. `cls.model_validate(data, context={"skip_validation": True})`), the model is constructed without validation.
        Any fields with a `Model` type will be constructed from their dict representation recursively.
        """
        context = info.context or {}
        try:
            return handler(value)
        except ValidationError:
            context = info.context or {}
            if context.get("skip_validation", False):
                model = cls.model_construct(**value)
                if not model.model_fields_set:
                    raise
                for field_name in model.model_fields_set:
                    field_value = getattr(model, field_name)
                    cls.__pydantic_validator__.validate_assignment(
                        model,
                        field_name,
                        field_value,
                        strict=(info.config or {}).get("strict"),
                        context=context,
                    )
                return model
            else:
                raise

    @field_validator("*", mode="wrap")
    def _field_validator(
        cls, value: Any, handler: ValidatorFunctionWrapHandler, info: ValidationInfo
    ):
        """The default field validator of the model.

        When validation is called with a `skip_validation=True` context, the field is assigned without validation.
        If the field is a `Model` type, the model will be constructed from its dict representation recursively.
        """
        context = info.context or {}
        try:
            return handler(value)
        except ValidationError:
            context = info.context or {}
            if context.get("skip_validation", False):
                if info.field_name:
                    field_type = get_type_hints(cls).get(info.field_name, Any)
                    try:
                        config = (
                            ConfigDict(arbitrary_types_allowed=True, strict=False)
                            if not isclass(field_type)
                            or not issubclass(field_type, PydanticBaseModel)
                            else None
                        )
                        return TypeAdapter(field_type, config=config).validate_python(
                            value, strict=False, context=context
                        )
                    except ValidationError:
                        return value
            else:
                raise

    def to_dict(self) -> Dict[str, Any]:
        """Convert the model instance to dict."""
        return self.model_dump()

    def to_json(self) -> str:
        """Convert the model instance to a JSON-encoded string."""
        return self.model_dump_json()

    @classmethod
    def from_dict(cls, obj: dict) -> Self:
        """Create a model instance from a dict."""
        return cls.model_validate(obj)

    @classmethod
    def from_json(cls, json_data: str | bytes | bytearray) -> Self:
        """Create a model instance from a JSON-encoded string."""
        return cls.model_validate_json(json_data)


class _Model(BaseModel):
    """A simple model that allows all additional attributes."""

    model_config = ConfigDict(extra="allow")

    def __init__(self, /, **kwargs):
        super().__init__(**kwargs)
        for field in self.model_fields_set:  # pylint: disable=not-an-iterable
            setattr(
                self, field, _Model.__model_construct_recursive(getattr(self, field))
            )

    @classmethod
    def __model_construct_recursive(cls, obj: Any):
        if isinstance(obj, list):
            return [cls.__model_construct_recursive(inner) for inner in obj]
        elif isinstance(obj, dict):
            model = _Model(**obj)
            return model
        else:
            return obj


Primitive: TypeAlias = Union[
    str, bool, int, float, Decimal, bytes, datetime, date, object, None
]
Model: TypeAlias = TypeAliasType(  # type: ignore[valid-type]  #(https://github.com/python/mypy/issues/16614)
    "Model",
    Annotated[
        Union[List["Model"], "_Model", Primitive],  # type: ignore[misc,possible cyclic definition]
        "A basic model that acts like a `simpleNamespace`, or a collection over such models.",
    ],
)
