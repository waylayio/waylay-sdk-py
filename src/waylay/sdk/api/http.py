"""Aliases for the http client."""

from __future__ import annotations
from typing import (
    TypedDict,
    Mapping,
    Optional,
    Callable,
    Any,
    Union,
    Sequence,
    Tuple,
    List,
    IO,
)
from typing_extensions import Required  # >=3.11

import httpx._client as httpx_types
import httpx

AsyncClient = httpx.AsyncClient
Response = httpx.Response
Request = httpx.Request

HeaderTypes = Union[
    Mapping[str, str],
    Mapping[bytes, bytes],
    Sequence[Tuple[str, str]],
    Sequence[Tuple[bytes, bytes]],
]

PrimitiveData = Optional[Union[str, int, float, bool]]
QueryParamTypes = Union[
    Mapping[str, Union[PrimitiveData, Sequence[PrimitiveData]]],
    List[Tuple[str, PrimitiveData]],
    Tuple[Tuple[str, PrimitiveData], ...],
    str,
    bytes,
]

RequestFiles = httpx_types.RequestFiles
RequestContent = Union[httpx_types.RequestContent, IO[bytes]]
RequestData = httpx_types.RequestData


class HttpRequestArguments(TypedDict, total=False):
    """Call parameters supported in an httpx request."""

    method: Required[str]
    url: Required[httpx_types.URLTypes]
    content: httpx_types.RequestContent
    data: httpx_types.RequestData
    files: RequestFiles
    json: Any
    params: QueryParamTypes
    headers: HeaderTypes
    cookies: httpx_types.CookieTypes
    auth: httpx_types.AuthTypes
    follow_redirects: bool
    timeout: httpx_types.TimeoutTypes
    extensions: httpx_types.RequestExtensions


class HttpClientOptions(TypedDict, total=False):
    """Options passed to the httpx client."""

    auth: Optional[httpx_types.AuthTypes]  # explicit None do disable auth
    params: httpx_types.QueryParamTypes
    headers: HeaderTypes
    cookies: httpx_types.CookieTypes
    verify: httpx_types.VerifyTypes
    cert: httpx_types.CertTypes
    http1: bool
    http2: bool
    proxy: httpx_types.ProxyTypes
    proxies: httpx_types.ProxiesTypes
    mounts: Mapping[str, Optional[httpx_types.AsyncBaseTransport]]
    timeout: httpx_types.TimeoutTypes
    follow_redirects: bool
    limits: httpx_types.Limits
    max_redirects: int
    event_hooks: Mapping[str, list[Callable[..., Any]]]
    base_url: str
    transport: httpx_types.AsyncBaseTransport
    app: Callable[..., Any]
    trust_env: bool
    default_encoding: str | Callable[[bytes], str]
