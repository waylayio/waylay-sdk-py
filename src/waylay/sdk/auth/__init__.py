"""Waylay SDK authentication."""

from .model import (
    WaylayCredentials,
    ClientCredentials,
    ApplicationCredentials,
    TokenCredentials,
    NoCredentials,
    WaylayToken,
)
from .provider import WaylayTokenAuth
from .parse import parse_credentials

from .exceptions import AuthError

__all__ = [
    "WaylayCredentials",
    "ClientCredentials",
    "ApplicationCredentials",
    "TokenCredentials",
    "NoCredentials",
    "WaylayToken",
    "WaylayTokenAuth",
    "parse_credentials",
    "AuthError",
]
