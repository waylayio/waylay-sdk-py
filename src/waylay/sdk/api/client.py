"""API client."""

from __future__ import annotations
from typing import Mapping, Optional, Tuple, Union

from ..__version__ import __version__
from ..config import WaylayConfig

from .exceptions import SyncCtxMgtNotSupportedError
from .serialization import WithSerializationSupport
from .http import AsyncClient, HttpClientOptions, Request, Response

RESTTimeout = Union[
    Optional[float],
    Tuple[Optional[float], Optional[float], Optional[float], Optional[float]],
]


class ApiClient(WithSerializationSupport):
    """Generic API client for OpenAPI client library builds.

    OpenAPI generic API client. This client handles the client-server
    communication, and is invariant across implementations. Specifics of
    the methods and models for each application are generated from the
    OpenAPI templates.

    :param configuration: configuration object for this client

    """

    config: WaylayConfig
    http_options: HttpClientOptions
    base_url: str
    _http_client: Optional[AsyncClient]

    def __init__(
        self,
        config: WaylayConfig,
        http_options: Optional[HttpClientOptions | AsyncClient] = None,
    ) -> None:
        """Create an instance."""
        self.config = config
        self.http_options = {}
        self._http_client = None
        self.base_url = self.config.gateway_url
        if http_options:
            self.set_options(http_options)

    def set_options(self, http_options: HttpClientOptions | AsyncClient | None):
        """Update http options on not open client."""
        if not self.is_closed:
            raise AttributeError("Cannot set options on open client.")
        if isinstance(http_options, AsyncClient):
            self._http_client = http_options
        elif isinstance(http_options, Mapping):
            self.http_options = http_options

    def clone(self, http_options: Optional[HttpClientOptions] = None):
        """Clone api client without http client."""
        http_options = (
            self.http_options
            if http_options is None
            else {**self.http_options, **http_options}
        )
        return self.__class__(self.config, http_options)

    @property
    def http_client(self) -> AsyncClient:
        """Get (or open) a http client."""
        if self._http_client is None or self.is_closed:
            self._http_client = self._init_http_client()
            return self._http_client
        return self._http_client

    @property
    def is_closed(self) -> bool:
        """Check that there is no active http client."""
        if not self._http_client:
            return True
        if self._http_client.is_closed:
            # someone else closed this http client, remove it.
            self._http_client = None
            return True
        return False

    async def __aenter__(self):
        """Initialize the http client."""
        if not self.is_closed:
            raise ValueError("Http client already in use.")
        self._http_client = self._init_http_client()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Close the http client."""
        await self.aclose()

    def __enter__(self):
        """Dissallow sync context management for now."""
        raise SyncCtxMgtNotSupportedError(self)

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Disallow sync context management."""
        raise SyncCtxMgtNotSupportedError(self)

    def _init_http_client(self) -> AsyncClient:
        opts = self.http_options
        client = AsyncClient(**opts)
        base_url = opts.get("base_url", self.config.gateway_url)
        client.base_url = base_url  # type: ignore
        auth = opts.get("auth", self.config.auth)
        client.auth = auth  # type: ignore
        client.headers.update({"User-Agent": f"waylay-sdk/python/{__version__}"})
        return client

    async def aclose(self):
        """Close the client."""
        if not self.is_closed:
            await self._http_client.aclose()
            self._http_client = None

    async def _request(self, *args, **kwargs) -> Response:
        """Invoke a http request."""
        return await self.http_client.request(*args, **kwargs)

    async def send(
        self, request: Request, *, stream: bool = False, **kwargs
    ) -> Response:
        """Send an http request."""
        return await self.http_client.send(request, stream=stream, **kwargs)
