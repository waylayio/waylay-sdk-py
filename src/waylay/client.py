"""REST client for the Waylay Platform."""

__version__ = '0.0.0'

from typing import List, Optional


from waylay.api.api_client import ApiClient
from waylay.api.api_config import ApiConfig
from .exceptions import ConfigError

from waylay.services._base import WaylayService
from waylay.services._loader import RegistryService, registry_available

from .config import (
    WaylayConfig,
    TenantSettings,
    DEFAULT_PROFILE,
    SERVICE_KEY_GATEWAY,
    SERVICE_KEY_ACCOUNTS
)
from .auth import (
    WaylayCredentials,
    ClientCredentials,
    TokenCredentials,
)


class WaylayClient():
    """REST client for the Waylay Platform."""

    config: WaylayConfig
    api_client: ApiClient

    # services
    registry: 'RegistryService'
    # TODO, do we want to load the services dynamically like in previous waylay-py
    # + only register installed services
    # - lose type info

    @classmethod
    def from_profile(
        cls, profile: str = DEFAULT_PROFILE,
        *, interactive=True, gateway_url=None
    ):
        """Create a WaylayClient named profile.

        Uses credentials from environment variables or a locally stored
        configuration.

        """
        return cls(WaylayConfig.load(
            profile, interactive=interactive, gateway_url=gateway_url
        ))

    @classmethod
    def from_client_credentials(
        cls, api_key: str, api_secret: str, *,
        gateway_url=None, accounts_url=None,
        settings: Optional[TenantSettings] = None
    ):
        """Create a WaylayClient using the given client credentials."""
        credentials = ClientCredentials(
            api_key, api_secret, **_auth_urls(gateway_url, accounts_url, settings)
        )
        return cls.from_credentials(credentials, settings=settings)

    @classmethod
    def from_token(
        cls, token_string: str, *,
        gateway_url=None, accounts_url=None, settings: Optional[TenantSettings] = None
    ):
        """Create a WaylayClient using a waylay token."""
        credentials = TokenCredentials(
            token_string, **_auth_urls(gateway_url, accounts_url, settings)
        )
        return cls.from_credentials(credentials, settings=settings)

    @classmethod
    def from_credentials(
        cls, credentials: WaylayCredentials, settings: Optional[TenantSettings] = None
    ):
        """Create a WaylayClient using the given client credentials."""
        return cls(WaylayConfig(
            credentials,
            settings=settings
        ))

    def __init__(self, config: WaylayConfig):
        """Create a WaylayConfig instance."""
        self._services: List[WaylayService] = []
        self.config = config
        self.load_services(config)

    def __repr__(self):
        """Get a technical string representation of this instance."""
        return (
            f"<{self.__class__.__name__}("
            # f"services=[{','.join(list(srv_class.service_key for srv_class in self.services))}],"
            f"config={self.config}"
            ")>"
        )

    @property
    def services(self) -> List[WaylayService]:
        """Get the services that are available through this client."""
        return self._services

    def configure(self, config: WaylayConfig):
        """Update this client with the given configuration."""
        self.config = config
        for srv in self._services:
            srv.configure(ApiClient(ApiConfig(config)))

    def load_services(self, config: WaylayConfig):
        """Load all services that are installed."""
        self.api_client = ApiClient(ApiConfig(config))

        self.registry = RegistryService(self.api_client)
        if registry_available:
            self._services.append(self.registry)


def _auth_urls(gateway_url=None, accounts_url=None, settings: Optional[TenantSettings] = None):
    if settings:
        gateway_url = gateway_url or settings.get(SERVICE_KEY_GATEWAY)
        accounts_url = accounts_url or settings.get(SERVICE_KEY_ACCOUNTS)
    return {'gateway_url': gateway_url, 'accounts_url': accounts_url}
