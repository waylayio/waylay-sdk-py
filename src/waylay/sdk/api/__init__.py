"""Waylay Api Configuration."""

# export api classes
from .client import ApiClient, RESTTimeout
from .http import (
    AsyncClient,
    HttpClientOptions,
    HeaderTypes,
    QueryParamTypes,
    Request,
    Response,
    RequestFiles,
    RequestContent,
    RequestData,
)
from .exceptions import ApiError, ApiValueError
from ._models import Model, Primitive

__all__ = [
    "ApiClient",
    "RESTTimeout",
    "AsyncClient",
    "HttpClientOptions",
    "HeaderTypes",
    "QueryParamTypes",
    "Request",
    "Response",
    "RequestFiles",
    "RequestContent",
    "RequestData",
    "ApiError",
    "ApiValueError",
    "Model",
    "Primitive",
]
