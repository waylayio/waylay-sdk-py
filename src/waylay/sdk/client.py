"""REST client for the Waylay Platform."""

from __future__ import annotations
from typing import TYPE_CHECKING, Optional

from .plugin.client import WithServicesAndTools
from .config.client import WaylayConfig, WithConfig, HttpClientOptions
from .api import ApiClient
from .api.http import AsyncClient


if TYPE_CHECKING:
    from .plugin.type_stubs import BrokerService, RegistryService


class WaylayClient(WithConfig, WithServicesAndTools):
    """REST client for the Waylay Platform."""

    registry: "RegistryService"
    broker: "BrokerService"

    def __init__(
        self,
        config: WaylayConfig | ApiClient,
        /,
        options: Optional[HttpClientOptions | AsyncClient] = None,
    ):
        """Create a WaylayConfig instance."""
        client_config: WaylayConfig
        api_client: ApiClient
        if isinstance(config, ApiClient):
            api_client = config
            client_config = api_client.config
            if options:
                api_client.set_options(options)
        else:
            client_config = config
            api_client = ApiClient(client_config, options)
        WithConfig.__init__(self, client_config)
        WithServicesAndTools.__init__(self, api_client)

    def __repr__(self):
        """Get a technical string representation of this instance."""
        return (
            f"<{self.__class__.__name__}("
            f"services=[{','.join(self._services.keys())}],"
            f"tools=[{','.join(self._tools.keys())}],"
            f"config={self.config}"
            ")>"
        )
