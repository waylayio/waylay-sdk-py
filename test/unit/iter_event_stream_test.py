import json
from asyncio import sleep
from collections.abc import AsyncIterator

import httpx
import pytest
from sse_starlette.sse import EventSourceResponse as EventSourceStarletteResponse
from starlette.applications import Starlette
from starlette.requests import Request as StarletteRequest
from starlette.routing import Route as StarletteRoute

from waylay.sdk.api._models import BaseModel as WaylayBaseModel
from waylay.sdk.api.client import ApiClient
from waylay.sdk.api.http import HttpClientOptions
from waylay.sdk.auth.model import NoCredentials
from waylay.sdk.config.model import WaylayConfig


@pytest.fixture(name="events", scope="session")
def _events_fixture() -> list:
    return [
        {"event": "started", "data": json.dumps({"type": "build"})},
        {"event": "completed", "data": json.dumps({"type": "build", "returnvalue": 0})},
    ]


@pytest.fixture(name="sse_app", scope="session")
def _sse_app_fixture(events: list):
    async def event_generator(events: list) -> AsyncIterator:
        for event in events:
            await sleep(0.3)
            yield event

    async def event_stream_endpoint(request: StarletteRequest):
        generator = event_generator(events)
        return EventSourceStarletteResponse(generator)

    app = Starlette(routes=[StarletteRoute("/", endpoint=event_stream_endpoint)])
    yield app
    del app


@pytest.fixture(name="app_transport", scope="session")
async def _app_transport_fixture(sse_app):
    transport = httpx.ASGITransport(app=sse_app)
    yield transport
    await transport.aclose()


@pytest.fixture(name="config", scope="session")
def _config_fixture():
    yield WaylayConfig(NoCredentials())


@pytest.fixture(name="client", scope="session")
async def _client_fixture(config, app_transport):
    http_opts: HttpClientOptions = {
        "transport": app_transport,
        "base_url": "http://test",
        "auth": None,
    }
    yield ApiClient(config, http_opts)


class EventData(WaylayBaseModel):
    """Event data model."""

    type: str
    returnvalue: int


class EventModel(WaylayBaseModel):
    """Event model."""

    event: str
    data: EventData


CASES = [
    (
        "dict_events",
        {
            "response_type": dict,
            "select_path": "",
        },
    ),
    (
        "model_events",
        {
            "response_type": EventModel,
            "select_path": "",
        },
    ),
    (
        "dict_events_select_path",
        {
            "response_type": dict,
            "select_path": "data",
        },
    ),
    (
        "model_events_select_path",
        {
            "response_type": EventData,
            "select_path": "data",
        },
    ),
]


def list_to_async_iterator(lst):
    for item in lst:
        yield item


@pytest.mark.parametrize(
    "test_input",
    [c[1] for c in CASES],
    ids=[c[0] for c in CASES],
)
@pytest.mark.asyncio(scope="session")
async def test_iter_event_stream(test_input, client, snapshot):
    async with client:
        event_stream = await client.request(
            "GET",
            "/",
            response_type=test_input.get("response_type"),
            select_path=test_input.get("select_path"),
            stream=True,
        )
        async for event in event_stream:
            assert event == snapshot()
