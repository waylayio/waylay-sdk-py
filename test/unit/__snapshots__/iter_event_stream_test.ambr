# serializer version: 1
# name: test_iter_event_stream[dict_events]
  dict({
    'data': dict({
      'type': 'build',
    }),
    'event': 'started',
  })
# ---
# name: test_iter_event_stream[dict_events].1
  dict({
    'data': dict({
      'returnvalue': 0,
      'type': 'build',
    }),
    'event': 'completed',
  })
# ---
# name: test_iter_event_stream[dict_events_select_path]
  dict({
    'type': 'build',
  })
# ---
# name: test_iter_event_stream[dict_events_select_path].1
  dict({
    'returnvalue': 0,
    'type': 'build',
  })
# ---
# name: test_iter_event_stream[model_events]
  EventModel(event='started', data=EventData(type='build', returnvalue=None))
# ---
# name: test_iter_event_stream[model_events].1
  EventModel(event='completed', data=EventData(type='build', returnvalue=0))
# ---
# name: test_iter_event_stream[model_events_select_path]
  EventData(type='build', returnvalue=None)
# ---
# name: test_iter_event_stream[model_events_select_path].1
  EventData(type='build', returnvalue=0)
# ---
